sql/blocks/blocks.go

```diff
@@ -0,0 +1,122 @@
+	bytes, err := codec.Encode(block.InnerBlock)
+	if err != nil {
+		return fmt.Errorf("encode %w", err)
+	}
+	if _, err := db.Exec("insert into blocks (id, layer, block) values (?1, ?2, ?3);",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, block.ID().Bytes())
+			stmt.BindInt64(2, int64(block.LayerIndex.Value))
+			stmt.BindBytes(3, bytes) // this is actually should encode block
+		}, nil); err != nil {
+		return fmt.Errorf("insert %s: %w", block.ID(), err)
+	}
+	return nil
+}
+
+// Has a block in the database.
+func Has(db sql.Executor, id types.BlockID) (bool, error) {
+	rows, err := db.Exec("select 1 from blocks where id = ?1;",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, nil,
+	)
+	if err != nil {
+		return false, fmt.Errorf("has ballot %s: %w", id, err)
+	}
+	return rows > 0, nil
+}
+
+// Get block with id from database.
+func Get(db sql.Executor, id types.BlockID) (rst *types.Block, err error) {
+	if rows, err := db.Exec("select block from blocks where id = ?1;", func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}, func(stmt *sql.Statement) bool {
+		rst, err = decodeBlock(stmt.ColumnReader(0), id)
+		return true
+	}); err != nil {
+		return nil, fmt.Errorf("get %s: %w", id, err)
+	} else if rows == 0 {
+		return nil, fmt.Errorf("%w block %s", sql.ErrNotFound, id)
+	}
+	return rst, err
+}
+
+// SetValid updates verified status for a block.
+func SetValid(db sql.Executor, id types.BlockID) error {
+	return setValidity(db, id, 1)
+}
+
+// SetInvalid updates blocks to an invalid status.
+func SetInvalid(db sql.Executor, id types.BlockID) error {
+	return setValidity(db, id, -1)
+}
+
+func setValidity(db sql.Executor, id types.BlockID, validity int8) error {
+	if rows, err := db.Exec(`insert into blocks (id, validity) values (?1, ?2) 
+	on conflict(id) do update set validity=?2 returning id;`, func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+		stmt.BindInt64(2, int64(validity))
+	}, nil); err != nil {
+		return fmt.Errorf("update validity %s: %w", id, err)
+	} else if rows == 0 {
+		return fmt.Errorf("%w block for update %s", sql.ErrNotFound, id)
+	}
+	return nil
+}
+
+// IsValid returns true if block is verified.
+func IsValid(db sql.Executor, id types.BlockID) (rst bool, err error) {
+	if rows, err := db.Exec("select validity from blocks where id = ?1;", func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}, func(stmt *sql.Statement) bool {
+		if stmt.ColumnInt(0) == 0 {
+			err = fmt.Errorf("%w block %s is undecided", sql.ErrNotFound, id)
+			return false
+		}
+		rst = stmt.ColumnInt(0) == 1
+		return true
+	}); err != nil {
+		return false, fmt.Errorf("select verified %s: %w", id, err)
+	} else if rows == 0 {
+		return false, fmt.Errorf("%w block %s is not in the database", sql.ErrNotFound, err)
+	}
+	return rst, err
+}
+
+// LayerIDs returns list of block ids in the layer.
+func LayerIDs(db sql.Executor, lid types.LayerID) ([]types.BlockID, error) {
```

Comment:
the name implies returning a list of LayerID. call this LayerBlockIDs?

Reply:
blocks is a name of the module, maybe IDsInLayer